### DISCUSSION TITLE
Enhancement - Sensors for power plugs should use "long-term statistics"-sensor to allow energy monitoring in Home Assistant

### DISCUSSION THREAD

**Discussion by titiviking** [arendst/Tasmota#12845] [Ideas]
### PROBLEM DESCRIPTION

HA 2021.8 introduces Energy monitoring. This is based on sensors that are setup for long-term statistics.
Today the Tasmota sensors are not setup as such & thererfore cannot be added to the HA Energy Monitoring.

https://developers.home-assistant.io/docs/core/entity/sensor#long-term-statistics.
https://developers.home-assistant.io/docs/core/entity/sensor/#available-device-classes
https://www.home-assistant.io/blog/2021/08/04/home-energy-management/

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ ] Tasmota binary firmware version number used: _____
  - [ ] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
What changes do you envision done in Tasmota for this? Skimming the first link you posted makes it look like a question of what configuration is done in Home Assistant, only.

**Comment by ascillato** [comment]
In Home Assistant, you should go to CONFIGURATION, then ENERGY and there add your devices that you want to use in the new ENERGY TAB of Home Assistant.

From Tasmota side, there is nothing extra that has to be done AFAIK.

**Reply by ebcjaa** [reply]
I tried this using Tasmota integration in Home Assistant. In order to add a Energy device it has to show up in drop down list.
My Sonoff POW R2 shows up, but my scripted modbus interface that also has kWh values does not (I am not surprised)
Anyhow the only difference in home assistant core_entity_ registry is that the POW has capabilities  set;
"capabilities": {
                    "state_class": "measurement"
                },
It is only POW "Energy Total" that has capabilities, not energy today and energy yesterday.  
i.e. there must be a something from the Tasmota side that is done differently.

**Reply by ascillato2** [reply]
You are right. Changing this from Q&A to Ideas

**Reply by sfromis** [reply]
Chances of something being implemented would improve if you can point out very specifically what change to Tasmota could make it work, instead of the vague "something".

**Reply by ebcjaa** [reply]
I don't know what specifically trigger the Tasmota integration in home assistant to  set  "capabilities"
I have only tested  Sonoff POW R2 w/ Tasmota 9.5.0 where "Energy Total" can be used with the new home assistant "Energy" feature.
Apparently according to the problem description  "power plugs" does not.   
I have no idea what needs to be changed to make it work. Hence the vague "something".
Maybe someone with better knowledge than me has.

**Reply by blakadder** [reply]
@emontnemery can you confirm or deny this is an issue?

**Reply by MaxWinterstein** [reply]
Just dropping my two cents here:

I ran against the same as @ebcjaa: 
All my Gosund plugs work perfectly for the new 'energy dashboard' feature. 
Some ESP8266 with additional IR Reader for Smart Meter shows fine in the integration but is missing the needed properties.

Adding them via `customization` like this: (taken from `customization.yml`

```yaml
sensor.stromzaehler_wohnung_power_total:
  unit_of_measurement: kWh
  device_class: energy
  state_class: measurement
  last_reset: '2021-06-01T00:00:00+00:00'
```

Did the job just fine.

**Comment by emontnemery** [comment]
Home Assistant 2021.8 requires the following for energy meters to show up:
```yaml
  unit_of_measurement: kWh
  device_class: energy
  state_class: measurement
  last_reset: <a valid timestamp>
```

The purpose of the `last_reset` attribute is to detect a new meter cycle.

Starting with Home Assistant 2021.9, it's no longer necessary to set the `last_reset` attribute for energy meters which can only count upwards, a new meter cycle will be detected automatically if the value decreases.
The correct configuration starting with Home Assistant 2021.9 for that kind of meter will be:

```yaml
  unit_of_measurement: kWh
  device_class: energy
  state_class: total_increasing
```

What's the recommendation for the Tasmota integration, should it set the `last_reset` attribute or is that not needed?
The integration currently maps `TotalStartTime` to `last_reset`.

Home Assistant developer blog post (in progress): https://github.com/home-assistant/developers.home-assistant/pull/1023
Home Assistant core PR: https://github.com/home-assistant/core/pull/54523

**Reply by mrandt** [reply]
I'd like to second this feature request. 

I use Tasmota as an interface to my electricity meter. Running on an ESP8266, it reads both current energy consumption and total (lifetime) consumption from an Iskra MT175 meter via an optical interface using SML (smart meter language) protocol. It submits the readings to Homeassistant via MQTT.

In Homeassistant, the device is auto detected and configured using standard Tasmota integration.

I am using the "customization" feature of Homeassistant to overwrite the necessary attributes:

![grafik](https://user-images.githubusercontent.com/10500490/142852968-04ef23ec-53ff-42fa-b81d-fc938a4d7192.png)

It would be great, if the integration would set the necessary attributes automatically once a SML device is discovered.

**Reply by jacauc** [reply]
This is exactly what is required. I'm not sure if that needs to be done in the Tasmota integration or in the Tasmota firmware itself.
Without proper state class and device class, the long term statistics is not tracked.

See: https://www.home-assistant.io/more-info/statistics/

How did you do customization like the above?

I intend to use statistics graph to show real-time power usage of a tasmota POW device. Similar to how I can do this with another ESPhome Glow device in the below example.

For this, I believe the sensor.<sensorname>_energy_power needs to be defined with:

    unit_of_measurement: 'W'
    state_class: measurement
    device_class: power



![image](https://user-images.githubusercontent.com/18201320/145147456-122d76f5-a117-43af-9323-4af1c73a9d76.png)

**Reply by mrandt** [reply]
I have enabled customization in configuration.yaml as described here:
https://www.home-assistant.io/docs/configuration/customizing-devices/

Once you do that, you get a new "Customizations" Menu in "Configuration" Menu:

![grafik](https://user-images.githubusercontent.com/10500490/145207020-fcee3079-87d8-47f9-993b-75c9e1cce1f9.png)

This allows you to overwrite attributes per entity via the GUI - that is also what my screenshot in https://github.com/arendst/Tasmota/discussions/12845#discussioncomment-1680830 shows.

Basically, the customizations overwrite the attributes from the integrations or other config files.

Bottomline: I think Tasmota integration needs to be changed to set these attributes by default but the customization is a workaround.

**Reply by emontnemery** [reply]
Fix for the Tasmota integration here: https://github.com/home-assistant/core/pull/61236, I think this sets all numeric sensors to `measurement`, please let me know if i missed some

**Reply by jacauc** [reply]
> Fix for the Tasmota integration here: [home-assistant/core#61236](https://github.com/home-assistant/core/pull/61236), I think this sets all numeric sensors to `measurement`, please let me know if i missed some

Amazing thank you, how can I test it? Do I wait for the next beta release?

**Reply by jacauc** [reply]
> I have enabled customization in configuration.yaml as described here: https://www.home-assistant.io/docs/configuration/customizing-devices/
> 
> Once you do that, you get a new "Customizations" Menu in "Configuration" Menu:
> 
> ![grafik](https://user-images.githubusercontent.com/10500490/145207020-fcee3079-87d8-47f9-993b-75c9e1cce1f9.png)
> 
> This allows you to overwrite attributes per entity via the GUI - that is also what my screenshot in [#12845 (reply in thread)](https://github.com/arendst/Tasmota/discussions/12845#discussioncomment-1680830) shows.
> 
> Basically, the customizations overwrite the attributes from the integrations or other config files.
> 
> Bottomline: I think Tasmota integration needs to be changed to set these attributes by default but the customization is a workaround.

Using the latest 2021.12 branch of HA and the whole menu was re-arranged - can't find that anymore. Managed to do it with yaml though. Thank you!

**Reply by emontnemery** [reply]
> Amazing thank you, how can I test it? Do I wait for the next beta release?

Yeah, it should be included in the next beta.

> Using the latest 2021.12 branch of HA and the whole menu was re-arranged - can't find that anymore. Managed to do it with yaml though.

The customize feature is deprecated because it doesn't quite work as users often expected; the customizations are applied as an override of the state so the sensor itself doesn't know it has been customized.

**Reply by kongo09** [reply]
> I am using the "customization" feature of Homeassistant

The customization feature has been deprecated from the UI: https://www.home-assistant.io/blog/2021/12/11/release-202112/

**Comment by titiviking** [comment]
for release 2021.9 some additional changes have been implemented:
https://developers.home-assistant.io/blog/2021/08/16/state_class_total/

**Comment by ahmaddxb** [comment]
Would be nice if this was done in auto discovery with so19 1

**Reply by ascillato** [reply]
SO19 1 is being deprecated and no longer under development.

It is replaced by the Native Home Assistant Autodiscovery with SO19 0

**Reply by blakadder** [reply]
Use Tasmota integration, MQTT discovery is deprecated as noted in the documentation

**Comment by macf0x** [comment]
Just going to put it out there, I struggled to get my Tasmota Plugs with power metering  and after setting the customizations to ensure the device_class and state_class matched the requirements, they still refused to display.

After comparing a couple of devices that were displaying, the common factor was the entity ID ended in _kwh.

I have seen plenty of examples where it was working without this '_kwh', but this worked for me.

Perhaps the tasmota integration should be naming the entity id with kwh at the end?
